renderdoc/driver/vulkan/vk_pixelhistory.cpp

```diff
@@ -0,0 +1,607 @@
+  VkImage colorImage;
+  VkImageView colorImageView;
+  VkImage stencilImage;
+  VkImageView stencilImageView;
+  VkDeviceMemory gpuMem;
+};
+
+struct EventInfo
+{
+  uint8_t premod[16];
+  uint8_t premod_depth[8];
+  uint8_t postmod[16];
+  uint8_t postmod_depth[8];
+  uint8_t shadout[16];
+  uint8_t fixed_stencil[8];
+  uint8_t original_stencil[8];
+};
+
+struct VulkanPixelHistoryCallback : public VulkanDrawcallCallback
+{
+  VulkanPixelHistoryCallback(WrappedVulkan *vk, uint32_t x, uint32_t y, VkImage image,
+                             VkFormat format, uint32_t sampleMask, VkQueryPool occlusionPool,
+                             VkImageView colorImageView, VkImageView stencilImageView,
+                             VkImage colorImage, VkImage stencilImage, VkBuffer dstBuffer,
+                             const std::vector<EventUsage> &events)
+      : m_pDriver(vk),
+        m_X(x),
+        m_Y(y),
+        m_Image(image),
+        m_Format(format),
+        m_DstBuffer(dstBuffer),
+        m_EventIndex(0),
+        m_SampleMask(sampleMask),
+        m_OcclusionPool(occlusionPool),
+        m_ColorImageView(colorImageView),
+        m_StencilImageView(stencilImageView),
+        m_ColorImage(colorImage),
+        m_StencilImage(stencilImage),
+        m_PrevState(vk, NULL)
+  {
+    m_pDriver->SetDrawcallCB(this);
+    for(size_t i = 0; i < events.size(); i++)
+      m_Events.insert(std::make_pair(events[i].eventId, events[i]));
+  }
+  ~VulkanPixelHistoryCallback() { m_pDriver->SetDrawcallCB(NULL); }
+  void CopyPixel(uint32_t eid, VkCommandBuffer cmd, uint32_t offset)
+  {
+    CopyPixelParams colourCopyParams = {};
+    colourCopyParams.multisampled = false;    // TODO: multisampled
+    colourCopyParams.srcImage = m_Image;
+    colourCopyParams.srcImageLayout =
+        VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;    // TODO: image layout
+    colourCopyParams.srcImageFormat = m_Format;
+    colourCopyParams.imageOffset = VkOffset3D{int32_t(m_X), int32_t(m_Y), 0};
+    colourCopyParams.dstBuffer = m_DstBuffer;
+
+    m_pDriver->GetDebugManager()->PixelHistoryCopyPixel(cmd, colourCopyParams, offset);
+
+    const DrawcallDescription *draw = m_pDriver->GetDrawcall(eid);
+    if(draw && draw->depthOut != ResourceId())
+    {
+      // The draw call had a depth image attachment.
+      ResourceId depthImage = m_pDriver->GetResourceManager()->GetLiveID(draw->depthOut);
+      VulkanCreationInfo::Image imginfo = m_pDriver->GetDebugManager()->GetImageInfo(depthImage);
+      CopyPixelParams depthCopyParams = colourCopyParams;
+      depthCopyParams.depthcopy = true;
+      depthCopyParams.srcImage =
+          (VkImage)m_pDriver->GetResourceManager()->GetCurrentResource(depthImage);
+      depthCopyParams.srcImageLayout = VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL;
+      depthCopyParams.srcImageFormat = imginfo.format;
+      m_pDriver->GetDebugManager()->PixelHistoryCopyPixel(cmd, depthCopyParams, offset + 16);
```

Comment:
This 16 offset is liable to bit-rot if the `EventInfo` struct changes, it'd be better to adjust the defintion of `EventInfo` so that colour and depth for premod/postmod are always stored together in a sub-struct, and then use `offsetof` on that sub-struct.

Reply:
Done, thanks!